### PR TITLE
Switch from elasticsearch to PostgreSQL text search

### LINKED ISSUE #293: Make tests not to rely on the network
When running the test suite in the local machine this is what I get:

```
➜  timeoverflow (develop) b rspec spec                                                                                                    ✱

Randomized with seed 64581
.....................................................................F.....................................*..........

Pending: (Failures listed here are expected and do not affect your suite's status)

  1) User.actives should list users with active members
     # No reason given
     # ./spec/models/user_spec.rb:59


Failures:

  1) OffersController GET #index (search) populates an array of offers
     Failure/Error: Offer.__elasticsearch__.create_index!(force: true)
     
     Faraday::ConnectionFailed:
       Failed to open TCP connection to localhost:9200 (Connection refused - connect(2) for "localhost" port 9200)
     # ./spec/controllers/offers_controller_spec.rb:38:in `block (3 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # Errno::ECONNREFUSED:
     #   Connection refused - connect(2) for "localhost" port 9200
     #   ./spec/controllers/offers_controller_spec.rb:38:in `block (3 levels) in <top (required)>'

Deprecation Warnings:

Requiring `rspec/autorun` when running RSpec via the `rspec` command is deprecated. Called from /home/pau/dev/timeoverflow/vendor/bundle/ruby/2.3.0/gems/activesupport-4.2.5.2/lib/active_support/dependencies.rb:274:in `require'.


If you need more of the backtrace for any of these deprecations to
identify where to make the necessary changes, you can configure
`config.raise_errors_for_deprecations!`, and it will turn the
deprecation warnings into errors, giving you the full backtrace.

1 deprecation warning total

Finished in 10.99 seconds (files took 4.4 seconds to load)
118 examples, 1 failure, 1 pending

Failed examples:

rspec ./spec/controllers/offers_controller_spec.rb:46 # OffersController GET #index (search) populates an array of offers

Randomized with seed 64581
```

We should fix the tests not to rely on the network. This makes the test suite brittle. We can consider taking a look at https://github.com/bblimke/webmock. It might the tool for this job.

### PR DESCRIPTION
**Pull request by enricostano**
### WAT
Elasticsearch is a dependency difficult to maintain and having to maintain a mirrored copy of the resources is also difficult, as you can see looking at the code deleted by this PR.

We already have PostgreSQL as a dependency and since our needs are quite simple we'll use the PostgreSQL built in [Full Text Search](https://www.postgresql.org/docs/current/textsearch-intro.html).

The gem that I'm proposing is https://github.com/Casecommons/pg_search

### Depends on
https://github.com/coopdevs/timeoverflow-provisioning/pull/62

### Extra
Closes #293

### CODE REVIEW

**Comment by sseerrggii** [comment]
I was playing with it :wink: 

:x: I noticed that you can't search part of a word. If I search "pequeña" I don't have Posts with "pequeñas" word inside, this a difference with  the usage with ES 

Worked fine: 

* Case insensitive :heavy_check_mark: 
* Accent insensitive :heavy_check_mark: 

I tested: 

## Search logged out: 

* https://staging.timeoverflow.org/offers?q=prueba
* https://staging.timeoverflow.org/inquiries?q=yoga

* https://staging.timeoverflow.org/offers?q=Prueba
* https://staging.timeoverflow.org/inquiries?q=yOga

* https://staging.timeoverflow.org/offers?q=Prüèba
* https://staging.timeoverflow.org/inquiries?q=yÓga

I also want to filter by organization https://staging.timeoverflow.org/offers?org=2&q=prueba but don't apply org filter only https://staging.timeoverflow.org/offers?org=2 https://staging.timeoverflow.org/offers?org=1 is filtering by organization. (this is the current behaviour with ES but not the expected one)

## Search logged in in org 1:

 * https://staging.timeoverflow.org/offers?q=prueba
* https://staging.timeoverflow.org/inquiries?q=yoga

* https://staging.timeoverflow.org/offers?q=Prueba
* https://staging.timeoverflow.org/inquiries?q=yOga

* https://staging.timeoverflow.org/offers?q=Prüèba
* https://staging.timeoverflow.org/inquiries?q=yÓga

## Search logged in in org 2:
Same URLs test, only to ensure that org filter is working, and it is :heavy_check_mark:

**Comment by markets** [comment]
@enricostano I think we can solve that issue by using one of the following settings: https://github.com/Casecommons/pg_search#searching-using-different-search-features (some of them need extra PostgreSQL extensions). I'd try with `:tsearch` first, and then (if no luck) with `:trigram` (smarter, but requires an extension).

**Comment by markets** [comment]
@sseerrggii @enricostano Pushed #467 with partial word search working

**Comment by markets** [comment]
@sseerrggii @enricostano I think we can leave the "Extra" requirements proposed by Sergi for another branch, so we can move on this one (which simplifies our infra a lot :ok_hand:). What do you think?

**Comment by markets** [comment]
@enricostano @sseerrggii updated with develop and fixed conflicts!

**Comment by sseerrggii** [comment]
:green_apple:  Tested :green_apple:

**Comment by enricostano** [comment]
I would like to test the performance of this new approach using a production db dump, with hundreds of posts to search from.

**Comment by sseerrggii** [comment]
Speed Test (time to load page in chrome console)

## Logged in, search "angles"
* With Elastic Search: **467m**
* PostgreSQL text search: **644ms**

## Logged in, search "las"
* With Elastic Search: **718ms**
* PostgreSQL text search: **566ms**

## Logged out, search "angles"
* With Elastic Search: **416ms**
* PostgreSQL text search: **2,16s**

## Logged out, search "las"
* With Elastic Search: **735ms**
* PostgreSQL text search: **2,11s**

**Comment by Morantron** [comment]
might be interesting to play with this https://www.postgresql.org/docs/9.5/textsearch-tables.html#TEXTSEARCH-TABLES-INDEX

**Comment by enricostano** [comment]
> might be interesting to play with this https://www.postgresql.org/docs/9.5/textsearch-tables.html#TEXTSEARCH-TABLES-INDEX

Yep, I just wanted to do some benchmark first.

Even if "logged out search" (hitting the whole Post table) is not used by users (@sseerrggii 's says... I would like some numbers instead), I would look into indexes in any case.

**Comment by enricostano** [comment]
Relevant blog posts in order to tune our config:

https://www.compose.com/articles/indexing-for-full-text-search-in-postgresql/
http://rachbelaid.com/postgres-full-text-search-is-good-enough/ (multi language)

**Comment by enricostano** [comment]
Basically in order to improve performance we need to:

1. store the `ts_vector` instead of generating it every time
2. create an index for the `ts_vector` field

File: app/controllers/members_controller.rb [hunk 1]
```diff
@@ -20,11 +20,6 @@ def toggle_manager
   def toggle_active
     find_member
     @member.toggle(:active).save!
-    if @member.active
-      @member.add_all_posts_to_index
-    else
-      @member.remove_all_posts_from_index
-    end
```
Comment (enricostano) [inline app/controllers/members_controller.rb]: @sseerrggii are we sure we don't need this?

Comment (enricostano) [reply app/controllers/members_controller.rb]: Maybe we're covered by https://github.com/coopdevs/timeoverflow/pull/465/files#diff-ab4e9273f7df75aa5004a42f757d6f51R7 but we should test both cases just to make sure:

1. setting a user as `active`, will her old posts show up now in search results?
2. setting a user ar `inactive`, will her posts be hidden in search results?

**Comment by enricostano** [comment]
Some more benchmarking:

1. https://gist.github.com/enricostano/beb71bfef8696c7bbface03eb31faf0c

File: app/controllers/members_controller.rb (hunk 1 above)
Comment (sseerrggii) [reply app/controllers/members_controller.rb]: Yes, le'ts test it. I'm not sure the behavior without Elastic Search but we need to show only Post from active Members

File: .travis.yml [hunk 2]
```diff
@@ -10,11 +10,7 @@ before_script:
   - curl -L https://codeclimate.com/downloads/test-reporter/test-reporter-latest-linux-amd64 > ./cc-test-reporter
   - chmod +x ./cc-test-reporter
   - ./cc-test-reporter before-build
-  # allow elasticsearch to be ready - https://docs.travis-ci.com/user/database-setup/#ElasticSearch
-  - sleep 10
```
Comment (fcsonline) [inline .travis.yml]: ❤️

Comment (enricostano) [reply .travis.yml]: :kissing_heart:

File: app/controllers/members_controller.rb (hunk 1 above)
Comment (enricostano) [reply app/controllers/members_controller.rb]: Done in https://github.com/coopdevs/timeoverflow/pull/465/commits/5d0ec9ee9e03c6b732cbc3d6a58d20d7653ba31b

**Comment by enricostano** [comment]
[Post.active.of_active_members.search_by_query('angles').explain](https://gist.github.com/enricostano/ca1be220017dfe7f6de2ef5894b2e0bc)

@sauloperez :point_up_2: I'll try to add a couple of indexes on `posts.active` and `members.active`

**Comment by enricostano** [comment]
I tried adding those indexes but since both columns are boolean PG never use them....

**Comment by enricostano** [comment]
@sseerrggii says that's tested enough :sweat_smile: 

Let's merge it!

**Comment by markets** [comment]
Nice job @enricostano and @sseerrggii :clap: this is a very good infra and project change!

**Pull request merged**